File: DTSparkCore/src/main/java/com/kevin/java/transformations/GroupByKey_RDD.java
package com.kevin.java.transformations;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;

/**
 * @author kevin
 * @version 1.0
 * @description     相同的Key分组并将其Value值聚合在一起
 * @createDate 2018/12/28
 */
public class GroupByKey_RDD {

    public static void main(String[] args) {

        // 1.创建SparkConf，设置作业名称和模式
        SparkConf conf = new SparkConf().setAppName("GroupByKey_RDD").setMaster("local");

        // 2.基于Sparkconf对象创建一个SparkContext上下文，它是通往集群的唯一通道，且在创建时会创建任务调度器
        JavaSparkContext sc = new JavaSparkContext(conf);

        // 3.将Tuple2的数据转为rdd
        JavaPairRDD<String, Integer> parallelizePairs = sc.parallelizePairs(Arrays.asList(
                new Tuple2<String,Integer>("a",100),
                new Tuple2<String,Integer>("b",200),
                new Tuple2<String,Integer>("b",100),
                new Tuple2<String,Integer>("c",200),
                new Tuple2<String,Integer>("d",100),
                new Tuple2<String,Integer>("d",300)
        ));

        // 4.将相同的Key分组并将其Value值聚合在一起
        JavaPairRDD<String, Iterable<Integer>> groupByKey = parallelizePairs.groupByKey();

        // 5.遍历数据
        groupByKey.foreach(new VoidFunction<Tuple2<String, Iterable<Integer>>>(){
            @Override
            public void call(Tuple2<String, Iterable<Integer>> s) throws Exception {
                System.out.println(s);
            }
        });

        // 6.关闭
        sc.close();

    }
}

File: DTSparkCore/src/main/scala/com/kevin/scala/transformations/GroupByKey_RDD.scala
package com.kevin.scala.transformations

import org.apache.spark.{SparkConf, SparkContext}

/**
  * 相同的Key分组并将其Value值聚合在一起
  */
object GroupByKey_RDD {

  def main(args: Array[String]): Unit = {
    // 1.创建SparkConf
    val conf = new SparkConf().setAppName("GroupByKey_RDD").setMaster("local")
    // 2.创建SparkContext
    val sc = new SparkContext(conf)
    // 3.创建tuple2类型的集合数据
    val list = List(("a",100),("b",200),("b",100),("c",100),("c",300),("d",100),("d",200))
    // 4.parallelize将集合转成rdd,groupByKey将相同的key分组并将其Value聚合在一起,foreach遍历数据
    sc.parallelize(list).groupByKey().foreach(println(_))
    // 5.关闭sc
    sc.stop()
  }

}
